### ISSUE TITLE
Qhull segmentation fault

### ISSUE THREAD

**Issue by stephane-caron** [scipy/scipy#6484]
Running the attached example [qhull_test.txt](https://github.com/scipy/scipy/files/413705/qhull_test.txt) on my machine yields a segmentation fault:

```
python qhull_test.txt
Scipy version: 0.18.0
*** Error in `python': free(): invalid pointer: 0x0000000002231968 ***
[1]    22465 abort (core dumped)  python qhull_test.txt
```

However, I noticed that using [pyhull](http://pythonhosted.org/pyhull/) rather than `scipy.spatial`, there is no segmentation fault:

```
$ ipython -i qhull_test.txt  # replaced ScipyHull by PyHull
Python 2.7.6 (default, Jun 22 2015, 17:58:13) 
[...]
qhull precision warning: 
The initial hull is narrow (cosine of min. angle is 1.0000000000000000).
Is the input lower dimensional (e.g., on a plane in 3-d)? 
[...]
QH6227 
qhull precision error: Only 3 facets remain.  Can not merge another
pair.
[...]
In [1]:  # no segfault, here is the prompt
```

I cannot joggle the input (as suggest in [this bug report](https://github.com/scipy/scipy/issues/4974)) as it yields terrible outputs on my problem. Also, I found that the default in pyhull (in the ConvexHull constructor) is `joggling=False`, so pyhull does not need joggling to avoid this segfault.

**Edit:** also, I found that adding the `Q0` option raises a `QhullError` rather than segfaulting. That'll be my quickfix for now...

**Comment by pv** [comment]
Probably a Qhull bug. pyhull runs in a separate process, so segfaults
may not be visible --- also, the process quits without deallocating all
memory, which needs to be done if you run in same process.

**Comment by stephane-caron** [comment]
OK, I just reported the issue qhull_bug@qhull.org.

**Comment by pv** [comment]
Simplified C test program: https://gist.github.com/pv/b10c82a91ffe20dc6a1ddebd4e44c0b2

**Comment by pv** [comment]
The use of `qh_freeqhull(qh, qh_ALL)` here may be incorrect, however --- the crash goes away with `qh_freeqhull(qh, !qh_ALL)` and this does not appear to cause memory leak. 
Interesting that this is the first data set reported to cause a crash.

**Comment by cbbarber** [comment]
Qhull 2019.1 will fix this problem.  Pauli, Many thanks for writing a standalone test.   I'm sorry it took so long to resolve.

It appears to be three bugs, two of which from this test case.   

1) The segfault appears to be due to a qh_freebuild bug fixed in 2019.0.1.  Since the error occurred while adding a new point, ridges of visible facets (to be deleted) could be double deleted.

 - global_r.c/qh_freebuild: do not delete unattached ridges, otherwise a double delete.  Should not occur
and/or
- global_r.c/qh_freebuild: free global sets first instead of last, otherwise dangling references

2) The newly discovered bug is rather serious.  It is triggered by a coplanar interior_point, typically for a narrow hull.   The problem was that qh_initialhull used qh_checkflipped just like qh_setfacetplane.  It fails if the facet is 'maybe' flipped (i.e., distance >= - qh.DISTround).  For qh_initialhull, it should be checking for 'clearly' flipped (i.e., distance > qh.DISTround). 

- poly2_r.c/qh_initialhull: reverse orientation if clearly flipped.  qh_ERRsingular if maybe flipped [S. Caron, P. Virtanen, D. Sterratt, others]

3) The end of qh_initialhull checks for coplanar qh.interior_point (aka, qh_ERRsingular).   It called qh_checkflipped with allerror==False.  This tested qh.interior_point against 0.0.   Because of (2), qh.interior_point had be forced to be less than 0.0 from each facet.   So this test passed, when it should have failed.

- poly2_r.c/qh_initialhull: test for coplanar qh.interior_point with 'allerror' (dist >= -qh.DISTround) [S. Caron, P. Virtanen, D. Sterratt, others]

Qhull 2019.1 reports a qh_ERRsingular bug since the initial simplex built from min and max coordinate points is coplanar with the interior point.   Option Qs (search all points) builds a convex hull as desired.   The selection of the initial simplex will hopefully be improved in a later release.

kelwyn-/d/bash/local/qhull/bin>  qhull <../working/qhull_test.txt T4z | head -100

qh_maxmin: found the max and min points (by dim): p571 p422 p422 p572
qh_distround: 2.6e-13, maxabs 3e+02, maxsumabs 5.3e+02, maxdistsum 4.2e+02
[QH3056]qh_detmaxoutside: MAXoutside 1.5e-12 from qh.max_outside  0, ONEmerge 1.3e-12, MINoutside 1e-12, DISTround 2.6e-13

# qh_checkflipped reports flipped, which qh_initialhull ignores since the distance is less than qh.DISTround

qh_initialhull: qh.interior_point p-2:     100.6   -86.14
[QH0019]qh_checkflipped: facet f1 flipped, allerror? 1, distance= -1.27897692437e-13 during p-1
qh_distplane: -1.27897692436818e-13 from p-2 to f1

# Qhull reports qh_ERRsingular

QH6154 Qhull precision error: Initial simplex is flat (facet 1 is coplanar with the interior point)
center point    100.6   -86.14
facet p571 p422 distance= -1.3e-13
facet p572 p422 distance= -1.6e-13
facet p572 p571 distance= -2.3e+02

# In this case, option 'Qs' finds a non-singular initial hull

qhull <../working/qhull_test.txt Qs

Convex hull of 744 points in 2-d:

  Number of vertices: 4
  Number of facets: 4
  Number of points processed: 4
  Number of hyperplanes created: 5
  Maximum distance of point above facet: 6.8e-13 (0.4x)

Qhull 2019.1 will be release later this month.  A pre-release (without all of these fixes) is available on github
    https://github.com/qhull/qhull/wiki

**Comment by pv** [comment]
Thank you, upgrading to 2019.1 (#10413) indeed fixes this issue.
